Add Basic masking to New instrument view

GuiMacielPereira opened the pull request:
### Description of work
This PR adds the bare-bones for masking in the new instrument view. Currently only one shape is supported (a circle) and I included a list widget that allows the user to pick between masks, combine them and finally export them (to workspace, file or burn the mask permanently into the workspace).

### To test:
Use the new widget to add a circle, apply several masks, choose some of the masks and finally export the masks into a workspace, save the masks to a file and burn the masks permanently into the workspace.

<!-- Include sufficient instructions for someone unfamiliar with the application to test.
Ok to refer back to instructions in the issue.
-->

<!-- REMEMBER:
- Add labels, milestones, etc.
- Ensure the base of this PR is correct (e.g. release-next or main)
- Add release notes in separate file as per ([guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)), or justify their absence:
  *This does not require release notes* because <fill in an explanation of why>
-->

<!--  GATEKEEPER: When squashing, remove the section from HERE...  -->
---

### Reviewer

**Your comments will be used as part of the gatekeeper process.** Comment clearly on what you have checked and tested during your review. Provide an audit trail for any changes requested.

As per the [review guidelines](http://developer.mantidproject.org/ReviewingAPullRequest.html):

- Is the code of an acceptable quality? ([Code standards](http://developer.mantidproject.org/Standards/)/[GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html))
- Has a thorough functional test been performed? Do the changes handle unexpected input/situations?
- Are appropriately scoped unit and/or system tests provided?
- Do the release notes conform to the [guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html) and describe the changes appropriately?
- Has the relevant (user and developer) documentation been added/updated?
- If the PR author isn’t in the `mantid-developers` or `mantid-contributors` teams, add a review comment `rerun ci` to authorize/rerun the CI

### Gatekeeper

As per the [gatekeeping guidelines](https://developer.mantidproject.org/Gatekeeping.html):

- Has a thorough first line review been conducted, including functional testing?
- At a high-level, is the code quality sufficient?
- Are the base, milestone and labels correct?
<!--  GATEKEEPER: ...To HERE  -->

GuiMacielPereira committed:
Basic cylinder widget]

GuiMacielPereira committed:
Disable Rotation of Cylinder

GuiMacielPereira committed:
Add callback to select detectors inside cylinder

GuiMacielPereira committed:
Read masked detectors as missing

First basic strucure to ignore masked detectors when present in the file

GuiMacielPereira committed:
Try basic masking functunality

Not yet working feature

GuiMacielPereira committed:
Small refactoring of projection options

Cleaned projections options by transfering state into the model.
Added a cache for projections so that they are only calculated once when
a new projection is selected.
Changed access to several properties so that they always check with
self.is_pickable before access (accounts for changes in masking)

GuiMacielPereira committed:
Cleanup picking and callbacks

GuiMacielPereira committed:
Rename plotter meshes

GuiMacielPereira committed:
Add mesh of dark grey points to mark masked detectors

GuiMacielPereira committed:
Add masking box and disable masking on callback

GuiMacielPereira committed:
Expand bounds not to cut off cylinder

GuiMacielPereira committed:
Update picked detectors upon masking

Submitting a mask on picked detectors should unpick them, since they'll
no longer be available for picking

GuiMacielPereira committed:
Cache camera positions

Adding widgets makes it that the default plotter.reset_camera() method
includes the bounds of the widget, which messes up the camera position.
Having a cache for the camera positions fixes this issue (although it
introduces a new bug where the axis of the cylinder widget get
stretched)

GuiMacielPereira committed:
Fix axis and overlay of cylinder widget

The way to finally achieve a 2D effect is to call
set_parallel_projection() and update the camera position after the
widget is turned on

GuiMacielPereira committed:
Add mask list widget

GuiMacielPereira committed:
Add save mask to workspace

GuiMacielPereira committed:
Add save mask to xml file

GuiMacielPereira committed:
Update plot upon setting a mask

GuiMacielPereira committed:
Add option to overwrite original workspace

GuiMacielPereira committed:
Add button for clearing masks

GuiMacielPereira committed:
Merge main

GuiMacielPereira committed:
Fix projection after merge

GuiMacielPereira committed:
Fix 3D camera reset

GuiMacielPereira committed:
Add automatic enabling/disabling of add widget button

GuiMacielPereira committed:
Merge main

GuiMacielPereira committed:
Fix crash when saving xml is cancelled

GuiMacielPereira committed:
Fix update of workspace when replacing or renaming

Needed because if a mask is burned into the original workspace, the
observer for workspace replacement is called and model needs to be
updated

GuiMacielPereira committed:
Fix presenter tests

GuiMacielPereira committed:
Fix model tests

Can simplify most of the unit tests by using a function that sets up a
mock workspace and calls it with a model setup

GuiMacielPereira committed:
Add unit tests

GuiMacielPereira committed:
Merge branch 'main' into add-shapes-to-iv

jclarkeSTFC committed:
Merge branch 'main' into add-shapes-to-iv

jclarkeSTFC reviewed (commented):
This look really good, the circle works well. Saving/exporting all seemed to work as well. Sometimes the circle doesn't appear though, e.g. with MARI and the attached workspace. I suspect it's getting drawn off screen somewhere so maybe its initial position needs setting.

[SXD_peaks_masked.nxs.txt](https://github.com/user-attachments/files/23821885/SXD_peaks_masked.nxs.txt)

GuiMacielPereira committed:
Apply review changes

Co-authored-by: James Clarke <james.clarke@stfc.ac.uk>

GuiMacielPereira committed:
Center circle at data centre and scale radius

Circle is now at the centre of the mesh and the radius is proportional
to the data diagonal

jclarkeSTFC reviewed (commented):
This looks good, problems fixed, works well, I like the circle tool and the projection. I'm calling in the rabbit to give it a once-over as well.

GuiMacielPereira committed:
Apply suggestions from code rabbit

jclarkeSTFC reviewed (dismissed):
This is great, thanks for the changes

GuiMacielPereira committed:
Merge main

GuiMacielPereira committed:
Finish changes from merge

GuiMacielPereira committed:
Suppress rendering during plotter update

Introduced to fix issues with wrong transforms being calculated and
unpleasant flahes that were happening because the plotter was being
updated for every mesh added to it.

Instead of many incremental changes to the plotter it's better to
disable rendering, do all of the changes and then make those changes
render those changes to the plotter, so they appear all in one go.

Co-authored-by: James Clarke <james.clarke@stfc.ac.uk>

GuiMacielPereira committed:
Update transform to consider both detector and masking meshes

jclarkeSTFC reviewed (approved):
Looks good, sorry about the merge conflicts!
